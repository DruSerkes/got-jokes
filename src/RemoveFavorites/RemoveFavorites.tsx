import * as React from 'react';
import { useState } from 'react';
import Modal from 'react-bootstrap/Modal';
import Button from 'react-bootstrap/Button';


interface RemoveFavoritesProps {
    viewingFavorites: boolean;
    haveFavorites: boolean;
    clearFavorites: () => void;
}

export const RemoveFavorites = ({ viewingFavorites, haveFavorites, clearFavorites }: RemoveFavoritesProps) => {
    const [showRemoveModal, setShowRemoveModal] = useState(false);
    const handleShowRemoveModal = () => setShowRemoveModal(true);
    const handleClose = () => setShowRemoveModal(false);
    const handleClearFavorites = () => {
        clearFavorites();
        setShowRemoveModal(false);
    };
    return (
        <>
            { viewingFavorites && haveFavorites &&
                (<Button
                    variant="warning"
                    onClick={handleShowRemoveModal}
                    className='my-4'>
                    Clear Favorites
                </Button>)
            }
            <Modal show={showRemoveModal} onHide={handleClose}>
                <Modal.Header closeButton>
                    <Modal.Title>Confirm</Modal.Title>
                </Modal.Header>
                <Modal.Body>
                    Are you sure you want to clear your favorites?
              <br />
                    <small className="text-warning">This action cannot be undone!!</small>
                </Modal.Body>
                <Modal.Footer>
                    <Button variant="outline-secondary" onClick={handleClose}>
                        Cancel
                          </Button>
                    <Button variant="danger" onClick={handleClearFavorites}>
                        Remove All
                          </Button>
                </Modal.Footer>
            </Modal>
        </>
    )
};