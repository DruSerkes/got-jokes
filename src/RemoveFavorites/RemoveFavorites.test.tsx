import * as React from 'react';
import { fireEvent, render, screen, waitForElementToBeRemoved } from '@testing-library/react';
import { RemoveFavorites } from './RemoveFavorites';

describe('RemoveFavorites tests', () => {
    const clearFavorites = jest.fn().mockImplementation(() => console.log('calling clearFavorites'));

    it('should render without breaking and match snapshot', () => {
        const { asFragment } = render(<RemoveFavorites viewingFavorites haveFavorites clearFavorites={clearFavorites} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should render a "Clear Favorites" button', () => {
        render(<RemoveFavorites viewingFavorites haveFavorites clearFavorites={clearFavorites} />);
        const removeBtn = screen.getByText('Clear Favorites');
        expect(removeBtn).toBeInTheDocument();
    });

    it('should render a modal when you click the button', () => {
        render(<RemoveFavorites viewingFavorites haveFavorites clearFavorites={clearFavorites} />);
        const removeBtn = screen.getByText('Clear Favorites');
        fireEvent.click(removeBtn);
        const removeModal = screen.getByText('Confirm');
        expect(removeModal).toBeInTheDocument();
    });

    it('should call clearFavorites and close the modal when you click the button', async () => {
        render(<RemoveFavorites viewingFavorites haveFavorites clearFavorites={clearFavorites} />);
        const removeBtn = screen.getByText('Clear Favorites');
        fireEvent.click(removeBtn);
        const removeAllBtn = screen.getByText('Remove All');
        expect(removeAllBtn).toBeInTheDocument();
        fireEvent.click(removeAllBtn)
        expect(clearFavorites).toHaveBeenCalledTimes(1);
        await waitForElementToBeRemoved(() => screen.queryByText('Confirm'));
        const removeModal = screen.queryByText('Confirm');
        expect(removeModal).not.toBeInTheDocument()
    });
})