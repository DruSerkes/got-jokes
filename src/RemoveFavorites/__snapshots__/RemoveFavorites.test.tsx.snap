// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RemoveFavorites tests should render without breaking and match snapshot 1`] = `
<DocumentFragment>
  <button
    class="my-4 mx-2 btn btn-warning"
    type="button"
  >
    Clear Favorites
  </button>
</DocumentFragment>
`;
